Incompatibility with ansi-wl-pprint-1.0.2

I encountered the following error when I tried to compile `logic-TPTP` (`v0.5.0.0` or `master` (19b25627cff915c8c5224395e3154832a0073bf3)) with `ansi-wl-pprint-1.0.2`

```
[ 4 of 11] Compiling Codec.TPTP.Pretty

/Users/sakai/src/logic-TPTP/Codec/TPTP/Pretty.hs:195:1: error:
    Duplicate instance declarations:
      instance [overlap ok] Pretty a => Pretty (Identity a)
        -- Defined at Codec/TPTP/Pretty.hs:195:1
      instance [safe] Pretty a => Pretty (Identity a)
        -- Defined in ‘prettyprinter-1.7.1:Prettyprinter.Internal’
```

If I comment out the line:

https://github.com/DanielSchuessler/logic-TPTP/blob/19b25627cff915c8c5224395e3154832a0073bf3/Codec/TPTP/Pretty.hs#L195

It resulted in the following errors:

```
[ 4 of 11] Compiling Codec.TPTP.Pretty

/Users/sakai/src/logic-TPTP/Codec/TPTP/Pretty.hs:96:11: error:
    • Couldn't match type ‘ann’
                     with ‘prettyprinter-ansi-terminal-1.1.3:Prettyprinter.Render.Terminal.Internal.AnsiStyle’
      Expected: prettyprinter-1.7.1:Prettyprinter.Internal.Doc ann
        Actual: Doc
      ‘ann’ is a rigid type variable bound by
        the type signature for:
          pretty :: forall ann.
                    TPTP_Input -> prettyprinter-1.7.1:Prettyprinter.Internal.Doc ann
        at Codec/TPTP/Pretty.hs:95:5-10
    • In the expression:
        (text) "Formula " <+> (dullwhite . text) "name:"
          </>
            vsep
              [(red . pretty) name <+> dullwhite (text "role:")
                 <+> (magenta . text . unrole) role,
               pretty formula, pretty annotations, empty]
      In an equation for ‘pretty’:
          pretty AFormula {..}
            = (text) "Formula " <+> (dullwhite . text) "name:"
                </>
                  vsep
                    [(red . pretty) name <+> dullwhite (text "role:")
                       <+> (magenta . text . unrole) role,
                     pretty formula, pretty annotations, ....]
      In the instance declaration for ‘Pretty TPTP_Input’
    • Relevant bindings include
        pretty :: TPTP_Input
                  -> prettyprinter-1.7.1:Prettyprinter.Internal.Doc ann
          (bound at Codec/TPTP/Pretty.hs:95:5)
   |
96 |         = (text) "Formula " <+> (dullwhite.text) "name:"
   |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...

/Users/sakai/src/logic-TPTP/Codec/TPTP/Pretty.hs:112:34: error:
    • Couldn't match type ‘ann’
                     with ‘prettyprinter-ansi-terminal-1.1.3:Prettyprinter.Render.Terminal.Internal.AnsiStyle’
      Expected: prettyprinter-1.7.1:Prettyprinter.Internal.Doc ann
        Actual: Doc
      ‘ann’ is a rigid type variable bound by
        the type signature for:
          prettyList :: forall ann.
                        [TPTP_Input] -> prettyprinter-1.7.1:Prettyprinter.Internal.Doc ann
        at Codec/TPTP/Pretty.hs:112:5-14
    • In the expression: pretty x <$> xs
      In the first argument of ‘foldr’, namely
        ‘(\ x xs -> pretty x <$> xs)’
      In the expression: foldr (\ x xs -> pretty x <$> xs) empty
    • Relevant bindings include
        xs :: prettyprinter-1.7.1:Prettyprinter.Internal.Doc ann
          (bound at Codec/TPTP/Pretty.hs:112:28)
        prettyList :: [TPTP_Input]
                      -> prettyprinter-1.7.1:Prettyprinter.Internal.Doc ann
          (bound at Codec/TPTP/Pretty.hs:112:5)
    |
112 |     prettyList = foldr (\x xs -> pretty x <$> xs) empty
    |                                  ^^^^^^^^^^^^^^^

/Users/sakai/src/logic-TPTP/Codec/TPTP/Pretty.hs:115:18: error:
    • Couldn't match type ‘ann’
                     with ‘prettyprinter-ansi-terminal-1.1.3:Prettyprinter.Render.Terminal.Internal.AnsiStyle’
      Expected: prettyprinter-1.7.1:Prettyprinter.Internal.Doc ann
        Actual: Doc
      ‘ann’ is a rigid type variable bound by
        the type signature for:
          pretty :: forall ann.
                    Quant -> prettyprinter-1.7.1:Prettyprinter.Internal.Doc ann
        at Codec/TPTP/Pretty.hs:115:5-10
    • In the expression: oper "∀"
      In an equation for ‘pretty’: pretty All = oper "∀"
      In the instance declaration for ‘Pretty Quant’
    • Relevant bindings include
        pretty :: Quant
                  -> prettyprinter-1.7.1:Prettyprinter.Internal.Doc ann
          (bound at Codec/TPTP/Pretty.hs:115:5)
    |
115 |     pretty All = oper "∀"
    |                  ^^^^^^^^

/Users/sakai/src/logic-TPTP/Codec/TPTP/Pretty.hs:132:11: error:
    • Couldn't match type ‘ann’
                     with ‘prettyprinter-ansi-terminal-1.1.3:Prettyprinter.Render.Terminal.Internal.AnsiStyle’
      Expected: prettyprinter-1.7.1:Prettyprinter.Internal.Doc ann
        Actual: Doc
      ‘ann’ is a rigid type variable bound by
        the type signature for:
          pretty :: forall ann.
                    WithEnclosing (Formula0 t f)
                    -> prettyprinter-1.7.1:Prettyprinter.Internal.Doc ann
        at Codec/TPTP/Pretty.hs:119:5-10
    • In the expression:
        maybeParens newEnclosing enclosing
          $ case formu of
              Quant q vars f
                -> pretty q
                     <+> brackets (hsep (punctuate comma (fmap pretty vars)))
                     <> dot </> pretty (wne f)
              (:~:) f -> oper "¬" <+> pretty (wne f)
              PredApp p [] -> psym p
              PredApp p args -> psym p <> prettyargs (fmap wne args)
              BinOp f1 op f2
                -> align $ sep [indent 0 $ pretty (wne f1), pretty op, ....]
              InfixPred f1 op f2
                -> align $ sep [indent 0 $ pretty (wne f1), pretty op, ....]
      In the expression:
        let
          newEnclosing
            = case formu of
                Quant _ _ _ -> ...
                PredApp _ _ -> ...
                (:~:) _ -> ...
                BinOp _ op _ -> ...
                InfixPred _ op _ -> ...
          wne = WithEnclosing newEnclosing
        in
          maybeParens newEnclosing enclosing
            $ case formu of
                Quant q vars f
                  -> pretty q
                       <+> brackets (hsep (punctuate comma (fmap pretty vars)))
                       <> dot </> pretty (wne f)
                (:~:) f -> oper "¬" <+> pretty (wne f)
                PredApp p [] -> psym p
                PredApp p args -> psym p <> prettyargs (fmap wne args)
                BinOp f1 op f2 -> align $ sep [indent 0 $ pretty (wne f1), ....]
                InfixPred f1 op f2
                  -> align $ sep [indent 0 $ pretty (wne f1), ....]
      In an equation for ‘pretty’:
          pretty (WithEnclosing enclosing formu)
            = let
                newEnclosing = ...
                wne = WithEnclosing newEnclosing
              in
                maybeParens newEnclosing enclosing
                  $ case formu of
                      Quant q vars f
                        -> pretty q
                             <+> brackets (hsep (punctuate comma (fmap pretty vars)))
                             <> dot </> pretty (wne f)
                      (:~:) f -> oper "¬" <+> pretty (wne f)
                      PredApp p [] -> psym p
                      PredApp p args -> psym p <> prettyargs (fmap wne args)
                      BinOp f1 op f2 -> align $ sep [...]
                      InfixPred f1 op f2 -> align $ sep [...]
    • Relevant bindings include
        pretty :: WithEnclosing (Formula0 t f)
                  -> prettyprinter-1.7.1:Prettyprinter.Internal.Doc ann
          (bound at Codec/TPTP/Pretty.hs:119:5)
    |
132 |           maybeParens newEnclosing enclosing $
    |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...

/Users/sakai/src/logic-TPTP/Codec/TPTP/Pretty.hs:156:20: error:
    • Couldn't match type ‘ann’
                     with ‘prettyprinter-ansi-terminal-1.1.3:Prettyprinter.Render.Terminal.Internal.AnsiStyle’
      Expected: prettyprinter-1.7.1:Prettyprinter.Internal.Doc ann
        Actual: Doc
      ‘ann’ is a rigid type variable bound by
        the type signature for:
          pretty :: forall ann.
                    BinOp -> prettyprinter-1.7.1:Prettyprinter.Internal.Doc ann
        at Codec/TPTP/Pretty.hs:155:5-10
    • In the expression: oper "⇔"
      In a case alternative: (:<=>:) -> oper "⇔"
      In the expression:
        case x of
          (:<=>:) -> oper "⇔"
          (:=>:) -> oper "⇒"
          (:<=:) -> oper "⇐"
          (:&:) -> oper "∧"
          (:|:) -> oper "∨"
          (:~&:) -> oper "NAND"
          (:~|:) -> oper "NOR"
          (:<~>:) -> oper "XOR"
    • Relevant bindings include
        pretty :: BinOp
                  -> prettyprinter-1.7.1:Prettyprinter.Internal.Doc ann
          (bound at Codec/TPTP/Pretty.hs:155:5)
    |
156 |         (:<=>:) -> oper "⇔"
    |                    ^^^^^^^^

/Users/sakai/src/logic-TPTP/Codec/TPTP/Pretty.hs:167:20: error:
    • Couldn't match type ‘ann’
                     with ‘prettyprinter-ansi-terminal-1.1.3:Prettyprinter.Render.Terminal.Internal.AnsiStyle’
      Expected: prettyprinter-1.7.1:Prettyprinter.Internal.Doc ann
        Actual: Doc
      ‘ann’ is a rigid type variable bound by
        the type signature for:
          pretty :: forall ann.
                    InfixPred -> prettyprinter-1.7.1:Prettyprinter.Internal.Doc ann
        at Codec/TPTP/Pretty.hs:166:5-10
    • In the expression: oper "="
      In a case alternative: (:=:) -> oper "="
      In the expression:
        case x of
          (:=:) -> oper "="
          (:!=:) -> oper "≠"
    • Relevant bindings include
        pretty :: InfixPred
                  -> prettyprinter-1.7.1:Prettyprinter.Internal.Doc ann
          (bound at Codec/TPTP/Pretty.hs:166:5)
    |
167 |         (:=:)   -> oper "="
    |                    ^^^^^^^^

/Users/sakai/src/logic-TPTP/Codec/TPTP/Pretty.hs:174:30: error:
    • Couldn't match type ‘ann’
                     with ‘prettyprinter-ansi-terminal-1.1.3:Prettyprinter.Render.Terminal.Internal.AnsiStyle’
      Expected: prettyprinter-1.7.1:Prettyprinter.Internal.Doc ann
        Actual: Doc
      ‘ann’ is a rigid type variable bound by
        the type signature for:
          pretty :: forall ann.
                    WithEnclosing (Term0 t)
                    -> prettyprinter-1.7.1:Prettyprinter.Internal.Doc ann
        at Codec/TPTP/Pretty.hs:171:5-10
    • In the expression: prational d
      In a case alternative: NumberLitTerm d -> prational d
      In the expression:
        case x of
          Var s -> pretty s
          NumberLitTerm d -> prational d
          DistinctObjectTerm s -> cyan (dquotes (text s))
          FunApp f [] -> fsym f
          FunApp f args
            -> fsym f <> prettyargs (fmap (WithEnclosing EnclNothing) args)
    • Relevant bindings include
        pretty :: WithEnclosing (Term0 t)
                  -> prettyprinter-1.7.1:Prettyprinter.Internal.Doc ann
          (bound at Codec/TPTP/Pretty.hs:171:5)
    |
174 |           NumberLitTerm d -> prational d
    |                              ^^^^^^^^^^^

/Users/sakai/src/logic-TPTP/Codec/TPTP/Pretty.hs:207:28: error:
    • Couldn't match type ‘ann’
                     with ‘prettyprinter-ansi-terminal-1.1.3:Prettyprinter.Render.Terminal.Internal.AnsiStyle’
      Expected: Doc -> prettyprinter-1.7.1:Prettyprinter.Internal.Doc ann
        Actual: Doc -> Doc
      ‘ann’ is a rigid type variable bound by
        the type signature for:
          pretty :: forall ann.
                    Annotations -> prettyprinter-1.7.1:Prettyprinter.Internal.Doc ann
        at Codec/TPTP/Pretty.hs:207:5-10
    • In the first argument of ‘(.)’, namely ‘dullwhite’
      In the first argument of ‘($)’, namely ‘dullwhite . text’
      In the expression: dullwhite . text $ "NoAnnotations"
    • Relevant bindings include
        pretty :: Annotations
                  -> prettyprinter-1.7.1:Prettyprinter.Internal.Doc ann
          (bound at Codec/TPTP/Pretty.hs:207:5)
    |
207 |     pretty NoAnnotations = dullwhite . text $ "NoAnnotations"
    |                            ^^^^^^^^^

/Users/sakai/src/logic-TPTP/Codec/TPTP/Pretty.hs:211:27: error:
    • Couldn't match type ‘ann’
                     with ‘prettyprinter-ansi-terminal-1.1.3:Prettyprinter.Render.Terminal.Internal.AnsiStyle’
      Expected: prettyprinter-1.7.1:Prettyprinter.Internal.Doc ann
        Actual: Doc
      ‘ann’ is a rigid type variable bound by
        the type signature for:
          pretty :: forall ann.
                    UsefulInfo -> prettyprinter-1.7.1:Prettyprinter.Internal.Doc ann
        at Codec/TPTP/Pretty.hs:211:5-10
    • In the expression: empty
      In an equation for ‘pretty’: pretty NoUsefulInfo = empty
      In the instance declaration for ‘Pretty UsefulInfo’
    • Relevant bindings include
        pretty :: UsefulInfo
                  -> prettyprinter-1.7.1:Prettyprinter.Internal.Doc ann
          (bound at Codec/TPTP/Pretty.hs:211:5)
    |
211 |     pretty NoUsefulInfo = empty
    |                           ^^^^^

/Users/sakai/src/logic-TPTP/Codec/TPTP/Pretty.hs:218:26: error:
    • Couldn't match type ‘ann’
                     with ‘prettyprinter-ansi-terminal-1.1.3:Prettyprinter.Render.Terminal.Internal.AnsiStyle’
      Expected: prettyprinter-1.7.1:Prettyprinter.Internal.Doc ann
        Actual: Doc
      ‘ann’ is a rigid type variable bound by
        the type signature for:
          pretty :: forall ann.
                    GData -> prettyprinter-1.7.1:Prettyprinter.Internal.Doc ann
        at Codec/TPTP/Pretty.hs:217:5-10
    • In the expression: prational x
      In an equation for ‘pretty’: pretty (GNumber x) = prational x
      In the instance declaration for ‘Pretty GData’
    • Relevant bindings include
        pretty :: GData
                  -> prettyprinter-1.7.1:Prettyprinter.Internal.Doc ann
          (bound at Codec/TPTP/Pretty.hs:217:5)
    |
218 |     pretty (GNumber x) = prational x
    |                          ^^^^^^^^^^^

/Users/sakai/src/logic-TPTP/Codec/TPTP/Pretty.hs:227:29: error:
    • Couldn't match type ‘ann’
                     with ‘prettyprinter-ansi-terminal-1.1.3:Prettyprinter.Render.Terminal.Internal.AnsiStyle’
      Expected: prettyprinter-1.7.1:Prettyprinter.Internal.Doc ann
        Actual: Doc
      ‘ann’ is a rigid type variable bound by
        the type signature for:
          pretty :: forall ann.
                    AtomicWord -> prettyprinter-1.7.1:Prettyprinter.Internal.Doc ann
        at Codec/TPTP/Pretty.hs:227:5-10
    • In the expression:
        (if isLowerWord x then text else squotes . text) x
      In an equation for ‘pretty’:
          pretty (AtomicWord x)
            = (if isLowerWord x then text else squotes . text) x
      In the instance declaration for ‘Pretty AtomicWord’
    • Relevant bindings include
        pretty :: AtomicWord
                  -> prettyprinter-1.7.1:Prettyprinter.Internal.Doc ann
          (bound at Codec/TPTP/Pretty.hs:227:5)
    |
227 |     pretty (AtomicWord x) = (if isLowerWord x then text else squotes.text) x
    |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

/Users/sakai/src/logic-TPTP/Codec/TPTP/Pretty.hs:231:29: error:
    • Couldn't match type ‘ann’
                     with ‘prettyprinter-ansi-terminal-1.1.3:Prettyprinter.Render.Terminal.Internal.AnsiStyle’
      Expected: prettyprinter-1.7.1:Prettyprinter.Internal.Doc ann
        Actual: Doc
      ‘ann’ is a rigid type variable bound by
        the type signature for:
          pretty :: forall ann.
                    GTerm -> prettyprinter-1.7.1:Prettyprinter.Internal.Doc ann
        at Codec/TPTP/Pretty.hs:230:5-10
    • In the expression: pretty x <+> oper ":" <+> pretty y
      In an equation for ‘pretty’:
          pretty (ColonSep x y) = pretty x <+> oper ":" <+> pretty y
      In the instance declaration for ‘Pretty GTerm’
    • Relevant bindings include
        pretty :: GTerm
                  -> prettyprinter-1.7.1:Prettyprinter.Internal.Doc ann
          (bound at Codec/TPTP/Pretty.hs:230:5)
    |
231 |     pretty (ColonSep x y) = pretty x <+> oper ":" <+> pretty y
    |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

/Users/sakai/src/logic-TPTP/Codec/TPTP/Pretty.hs:239:20: error:
    • Couldn't match type ‘ann’
                     with ‘prettyprinter-ansi-terminal-1.1.3:Prettyprinter.Render.Terminal.Internal.AnsiStyle’
      Expected: Doc -> prettyprinter-1.7.1:Prettyprinter.Internal.Doc ann
        Actual: Doc -> Doc
      ‘ann’ is a rigid type variable bound by
        the type signature for:
          pretty :: forall ann.
                    V -> prettyprinter-1.7.1:Prettyprinter.Internal.Doc ann
        at Codec/TPTP/Pretty.hs:239:5-10
    • In the first argument of ‘(.)’, namely ‘blue’
      In the first argument of ‘($)’, namely ‘blue . text’
      In the expression: blue . text $ x
    • Relevant bindings include
        pretty :: V -> prettyprinter-1.7.1:Prettyprinter.Internal.Doc ann
          (bound at Codec/TPTP/Pretty.hs:239:5)
    |
239 |     pretty (V x) = blue . text $ x
    |                    ^^^^
```

--- a/Codec/TPTP/Diff.hs
+++ b/Codec/TPTP/Diff.hs
@@ -19,11 +19,14 @@ import Data.Generics
 import Test.QuickCheck hiding((.&.))
 import Codec.TPTP.Base
 import Codec.TPTP.Pretty
-import Text.PrettyPrint.ANSI.Leijen
+import Prettyprinter
+import Prettyprinter.Render.Terminal
 import Control.Monad
 import Control.Monad.Identity
 
 
+text :: String -> Doc ann
+text = pretty
 
 
 data DiffResult d =
@@ -210,55 +213,55 @@ type F0Diff = DiffResult (Formula0 (T DiffResult) (F DiffResult))
 wildcard :: AtomicWord
 wildcard = "_"
 
-instance Pretty (WithEnclosing T0Diff) where
-    pretty = prettyHelper --(plugSubterms (fApp wildcard []))
+instance PrettyAnsi (WithEnclosing T0Diff) where
+    prettyAnsi = prettyHelper --(plugSubterms (fApp wildcard []))
 
-instance Pretty (WithEnclosing F0Diff) where
-    pretty = prettyHelper --(plugSubformulae (fApp wildcard []) (pApp wildcard []))
+instance PrettyAnsi (WithEnclosing F0Diff) where
+    prettyAnsi = prettyHelper --(plugSubformulae (fApp wildcard []) (pApp wildcard []))
 
 prettyHelper :: forall t.
-                (Pretty (WithEnclosing t)) =>
-                WithEnclosing (DiffResult t) -> Doc
+                (PrettyAnsi (WithEnclosing t)) =>
+                WithEnclosing (DiffResult t) -> Doc AnsiStyle
 prettyHelper (WithEnclosing _ d) =
         let
-            pwe = pretty . WithEnclosing EnclNothing
+            pwe = prettyAnsi . WithEnclosing EnclNothing
         in
           case d of
-                 DontCare -> dullwhite.pretty $ wildcard
-                 Same -> dullgreen.text $ "Same"
-                 SameHead x -> blue . encloseSep (text "{ SameHead ") (text "}") semi $
+                 DontCare -> annotate (colorDull White).pretty $ wildcard
+                 Same -> annotate (colorDull Green).text $ "Same"
+                 SameHead x -> annotate (color Blue) . encloseSep (text "{ SameHead ") (text "}") semi $
                               [  pwe x
                               ]
 
 
 
-                 Differ x y -> red . encloseSep (text "{ Differ ") rbrace (text "; ") $
+                 Differ x y -> annotate (color Red) . encloseSep (text "{ Differ ") rbrace (text "; ") $
 
                               [ pwe x
                               , pwe y
                               ]
 
-deriving instance Pretty (T DiffResult)
+deriving instance PrettyAnsi (T DiffResult)
 
-deriving instance Pretty (F DiffResult)
+deriving instance PrettyAnsi (F DiffResult)
 
-instance Pretty (WithEnclosing (T DiffResult)) where
-    pretty (WithEnclosing x (T y)) = pretty (WithEnclosing x y)
+instance PrettyAnsi (WithEnclosing (T DiffResult)) where
+    prettyAnsi (WithEnclosing x (T y)) = prettyAnsi (WithEnclosing x y)
 
-instance Pretty (WithEnclosing (F DiffResult)) where
-    pretty (WithEnclosing x (F y)) = pretty (WithEnclosing x y)
+instance PrettyAnsi (WithEnclosing (F DiffResult)) where
+    prettyAnsi (WithEnclosing x (F y)) = prettyAnsi (WithEnclosing x y)
 
-instance Pretty (T0Diff) where
-    pretty = pretty . WithEnclosing EnclNothing
+instance PrettyAnsi (T0Diff) where
+    prettyAnsi = prettyAnsi . WithEnclosing EnclNothing
 
-instance Pretty (F0Diff) where
-    pretty = pretty . WithEnclosing EnclNothing
+instance PrettyAnsi (F0Diff) where
+    prettyAnsi = prettyAnsi . WithEnclosing EnclNothing
 
--- instance Pretty (DiffResult (T DiffResult)) where
---     pretty = pretty . WithEnclosing EnclNothing . T
+-- instance PrettyAnsi (DiffResult (T DiffResult)) where
+--     prettyAnsi = prettyAnsi . WithEnclosing EnclNothing . T
 
--- instance Pretty (DiffResult (T DiffResult) (F DiffResult)) where
---     pretty = pretty . WithEnclosing EnclNothing . F
+-- instance PrettyAnsi (DiffResult (T DiffResult) (F DiffResult)) where
+--     prettyAnsi = prettyAnsi . WithEnclosing EnclNothing . F
 
 
 -- runFormulaDiff :: Formula
@@ -300,12 +303,14 @@ printSampleDiffs = do
   xs <- sample' diffGenF
   ys <- sample' diffGenF
 
-  let item x y = text (replicate 50 '=')
-                 <$> pretty x
-                 <$> text (replicate 40 '-')
-                 <$> pretty y
-                 <$> text (replicate 40 '-')
-                 <$> pretty (diffFormula x y)
+  let item x y = vsep
+                 [ text (replicate 50 '=')
+                 , prettyAnsi x
+                 , text (replicate 40 '-')
+                 , prettyAnsi y
+                 , text (replicate 40 '-')
+                 , prettyAnsi (diffFormula x y)
+                 ]
 
   mapM_ (putStrLn . prettySimple . uncurry item) (zip xs ys)
 

--- a/Codec/TPTP/Pretty.hs
+++ b/Codec/TPTP/Pretty.hs
@@ -8,7 +8,7 @@
 {-# OPTIONS -Wall -fno-warn-orphans #-}
 
 -- | Mainly just 'Pretty' instances
-module Codec.TPTP.Pretty(prettySimple,WithEnclosing(..),Enclosing(..)) where
+module Codec.TPTP.Pretty(PrettyAnsi(..),prettySimple,WithEnclosing(..),Enclosing(..)) where
 
 #if MIN_VERSION_base(4,8,0)
 import Prelude hiding ((<$>))
@@ -16,34 +16,50 @@ import Prelude hiding ((<$>))
 
 import Codec.TPTP.Base
 import Codec.TPTP.Export
-import Text.PrettyPrint.ANSI.Leijen
+import Prettyprinter
+import Prettyprinter.Render.Terminal
 import Data.Data
 import Control.Monad.Identity
 import Data.Ratio
+import qualified Data.Text.Lazy as TL
 
-oper :: String -> Doc
-oper = dullyellow . text
-psym :: forall a. (Pretty a) => a -> Doc
-psym = green . pretty
-fsym :: forall a. (Pretty a) => a -> Doc
-fsym = yellow . pretty
+class PrettyAnsi a where
+  prettyAnsi :: a -> Doc AnsiStyle
+  prettyAnsiList :: [a] -> Doc AnsiStyle
+  prettyAnsiList = align . list . map prettyAnsi
+
+instance PrettyAnsi (Doc AnsiStyle) where
+  prettyAnsi = id
+
+instance PrettyAnsi a => PrettyAnsi [a] where
+  prettyAnsi = prettyAnsiList
+
+oper :: String -> Doc AnsiStyle
+oper = annotate (colorDull Yellow) . text
+psym :: forall a. (Pretty a) => a -> Doc AnsiStyle
+psym = annotate (color Green) . pretty
+fsym :: forall a. (Pretty a) => a -> Doc AnsiStyle
+fsym = annotate (color Yellow) . pretty
+
+text :: String -> Doc ann
+text = pretty
 
 --wtext :: String -> Doc
---wtext = white . text
+--wtext = annotate (color White) . text
 
-prettyargs :: forall a. (Pretty a) => [a] -> Doc
+prettyargs :: forall a. (PrettyAnsi a) => [a] -> Doc AnsiStyle
 --prettyargs = encloseSep (wtext "(") (wtext ")") (wtext ",") . fmap pretty
-prettyargs = tupled . fmap pretty
--- prettyargs [] = white $ text "()"
+prettyargs = tupled . fmap prettyAnsi
+-- prettyargs [] = annotate (color White) $ text "()"
 -- prettyargs (x:xs) =
---     let pxs = fmap ((white (text ",") <+>) . pretty) xs in
+--     let pxs = fmap ((annotate (color White) (text ",") <+>) . pretty) xs in
 
---     align (fillSep $ [    white (text "(") <+> pretty x ]
+--     align (fillSep $ [    annotate (color White) (text "(") <+> pretty x ]
 --                     ++ pxs
---                     ++ [ white (text ")") ] )
+--                     ++ [ annotate (color White) (text ")") ] )
 
-prational :: Rational -> Doc
-prational q = (text.show.numerator) q <> char '/' <> (text.show.denominator) q
+prational :: Rational -> Doc ann
+prational q = (text.show.numerator) q <> pretty '/' <> (text.show.denominator) q
 
 -- | Carries information about the enclosing operation (for the purpose of printing stuff without parentheses if possible).
 data WithEnclosing a = WithEnclosing Enclosing a
@@ -87,36 +103,37 @@ needsParens inner outer =
                EnclNothing -> (-1000) -- if there's no enclosing operation, we don't need parentheses
 
 
-maybeParens :: Enclosing -> Enclosing -> Doc -> Doc
+maybeParens :: Enclosing -> Enclosing -> Doc ann -> Doc ann
 maybeParens inner outer | needsParens inner outer = parens
 maybeParens _ _ = id
 
-instance Pretty TPTP_Input where
-    pretty AFormula{..}
-        = (text) "Formula " <+> (dullwhite.text) "name:"
-          </>
-          vsep
-          [
-            (red.pretty) name <+> dullwhite (text "role:") <+> (magenta.text.unrole) role
-          , pretty formula
-          , pretty annotations
-          ,empty
+instance PrettyAnsi TPTP_Input where
+    prettyAnsi AFormula{..}
+        = fillSep
+          [ (text) "Formula " <+> (annotate (colorDull White).text) "name:"
+          , vsep
+            [
+              (annotate (color Red).pretty) name <+> annotate (colorDull White) (text "role:") <+> (annotate (color Magenta).text.unrole) role
+            , prettyAnsi formula
+            , prettyAnsi annotations
+            , mempty
+            ]
           ]
 
-    pretty (Comment str)
-        = magenta . string $ str
+    prettyAnsi (Comment str)
+        = annotate (color Magenta) . pretty $ str
 
-    pretty (Include f sel) = text "include" <> parens (squotes (text f) <> (case sel of { [] -> empty; _ -> comma <+> sep (punctuate comma (fmap pretty sel)) } ))
+    prettyAnsi (Include f sel) = text "include" <> parens (squotes (text f) <> (case sel of { [] -> mempty; _ -> comma <+> sep (punctuate comma (fmap pretty sel)) } ))
 
 
-    prettyList = foldr (\x xs -> pretty x <$> xs) empty
+    prettyAnsiList = foldr (\x xs -> vsep [prettyAnsi x, xs]) mempty
 
-instance Pretty Quant where
-    pretty All = oper "∀"
-    pretty Exists = oper "∃"
+instance PrettyAnsi Quant where
+    prettyAnsi All = oper "∀"
+    prettyAnsi Exists = oper "∃"
 
-instance (Pretty (WithEnclosing t), Pretty (WithEnclosing f)) => Pretty ((WithEnclosing (Formula0 t f))) where
-    pretty (WithEnclosing enclosing formu) =
+instance (PrettyAnsi (WithEnclosing t), PrettyAnsi (WithEnclosing f)) => PrettyAnsi ((WithEnclosing (Formula0 t f))) where
+    prettyAnsi (WithEnclosing enclosing formu) =
         let
             newEnclosing =
              case formu of
@@ -134,25 +151,27 @@ instance (Pretty (WithEnclosing t), Pretty (WithEnclosing f)) => Pretty ((WithEn
           case formu of
 
             Quant q vars f ->
-               pretty q
-               <+> brackets (hsep (punctuate comma (fmap pretty vars)))
-               <> dot
-               </> pretty (wne f)
+               fillSep
+               [ prettyAnsi q
+                 <+> brackets (hsep (punctuate comma (fmap prettyAnsi vars)))
+                 <> dot
+               , prettyAnsi (wne f)
+               ]
 
-            (:~:) f -> oper "¬" <+> pretty (wne f)
+            (:~:) f -> oper "¬" <+> prettyAnsi (wne f)
 
             PredApp p [] -> psym p
             PredApp p args -> psym p <> prettyargs (fmap wne args)
 
 
             BinOp f1 op f2 ->
-                align $ sep [indent 0 $ pretty (wne f1), pretty op, indent 0 $ pretty (wne f2)]
+                align $ sep [indent 0 $ prettyAnsi (wne f1), prettyAnsi op, indent 0 $ prettyAnsi (wne f2)]
 
             InfixPred f1 op f2 ->
-                align $ sep [indent 0 $ pretty (wne f1), pretty op, indent 0 $ pretty (wne f2)]
+                align $ sep [indent 0 $ prettyAnsi (wne f1), prettyAnsi op, indent 0 $ prettyAnsi (wne f2)]
 
-instance Pretty BinOp where
-    pretty x = case x of
+instance PrettyAnsi BinOp where
+    prettyAnsi x = case x of
         (:<=>:) -> oper "⇔"
         (:=>:)  -> oper "⇒"
         (:<=:)  -> oper "⇐"
@@ -162,78 +181,79 @@ instance Pretty BinOp where
         (:~|:)  -> oper "NOR"
         (:<~>:) -> oper "XOR"
 
-instance Pretty InfixPred where
-    pretty x = case x of
+instance PrettyAnsi InfixPred where
+    prettyAnsi x = case x of
         (:=:)   -> oper "="
         (:!=:)  -> oper "≠"
 
-instance (Pretty (WithEnclosing t)) => Pretty (WithEnclosing (Term0 t)) where
-    pretty (WithEnclosing _ x) =
+instance (PrettyAnsi (WithEnclosing t)) => PrettyAnsi (WithEnclosing (Term0 t)) where
+    prettyAnsi (WithEnclosing _ x) =
         case x of
-          Var s -> pretty s
+          Var s -> prettyAnsi s
           NumberLitTerm d -> prational d
-          DistinctObjectTerm s -> cyan (dquotes (text s))
+          DistinctObjectTerm s -> annotate (color Cyan) (dquotes (text s))
           FunApp f [] -> fsym f
           FunApp f args -> fsym f <> prettyargs (fmap (WithEnclosing EnclNothing) args)
 
 
 
-instance Pretty (Formula0 Term Formula) where
-    pretty = pretty . WithEnclosing EnclNothing . F . Identity
+instance PrettyAnsi (Formula0 Term Formula) where
+    prettyAnsi = prettyAnsi . WithEnclosing EnclNothing . F . Identity
 
-instance Pretty (Term0 Term) where
-    pretty = pretty . WithEnclosing EnclNothing . T . Identity
+instance PrettyAnsi (Term0 Term) where
+    prettyAnsi = prettyAnsi . WithEnclosing EnclNothing . T . Identity
 
-instance Pretty (WithEnclosing Formula) where
-    pretty (WithEnclosing x (F (Identity y))) = pretty (WithEnclosing x y)
+instance PrettyAnsi (WithEnclosing Formula) where
+    prettyAnsi (WithEnclosing x (F (Identity y))) = prettyAnsi (WithEnclosing x y)
 
-instance Pretty (WithEnclosing Term) where
-    pretty (WithEnclosing x (T (Identity y))) = pretty (WithEnclosing x y)
+instance PrettyAnsi (WithEnclosing Term) where
+    prettyAnsi (WithEnclosing x (T (Identity y))) = prettyAnsi (WithEnclosing x y)
 
-deriving instance Pretty Formula
-deriving instance Pretty Term
-deriving instance Pretty a => Pretty (Identity a)
+deriving instance PrettyAnsi Formula
+deriving instance PrettyAnsi Term
+deriving instance PrettyAnsi a => PrettyAnsi (Identity a)
 
--- instance (Pretty f, Pretty t) => Pretty (Formula0 t f) where
+-- instance (Pretty f, Pretty t) => PrettyAnsi (Formula0 t f) where
 --     pretty = pretty . WithEnclosing ""
 
 -- instance (Pretty t) => Pretty (Term0 t) where
 --     pretty = pretty . WithEnclosing ""
 
-prettySimple :: Pretty a => a -> String
-prettySimple x = displayS (renderPretty 0.9 80 (pretty x)) ""
-
-instance Pretty Annotations where
-    pretty NoAnnotations = dullwhite . text $ "NoAnnotations"
-    pretty (Annotations a b) = dullwhite (text "SourceInfo: ") <+> pretty a <+> pretty b
-
-instance Pretty UsefulInfo where
-    pretty NoUsefulInfo = empty
-    pretty (UsefulInfo x) = dullwhite (text "UsefulInfo: ") <+> pretty x
+prettySimple :: PrettyAnsi a => a -> String
+prettySimple x = TL.unpack $ renderLazy $ layoutSmart LayoutOptions{ layoutPageWidth = AvailablePerLine 80 0.9 } $ prettyAnsi x
 
 
+instance PrettyAnsi Annotations where
+    prettyAnsi NoAnnotations = annotate (colorDull White) . text $ "NoAnnotations"
+    prettyAnsi (Annotations a b) = annotate (colorDull White) (text "SourceInfo: ") <+> prettyAnsi a <+> prettyAnsi b
 
-instance Pretty GData where
-    pretty (GWord x) = pretty x
-    pretty (GNumber x) = prational x
-    pretty (GDistinctObject x) = cyan (dquotes (text x))
-    pretty (GApp x []) = fsym x
-    pretty (GApp x args) = fsym x <+> prettyargs args
-    pretty (GFormulaData s f) = text s <> align (parens (pretty f))
-    pretty (GFormulaTerm s t) = text s <> align (parens (pretty t))
-    pretty (GVar x) = pretty x
+instance PrettyAnsi UsefulInfo where
+    prettyAnsi NoUsefulInfo = mempty
+    prettyAnsi (UsefulInfo x) = annotate (colorDull White) (text "UsefulInfo: ") <+> prettyAnsi x
+
+
+
+instance PrettyAnsi GData where
+    prettyAnsi (GWord x) = pretty x
+    prettyAnsi (GNumber x) = prational x
+    prettyAnsi (GDistinctObject x) = annotate (color Cyan) (dquotes (text x))
+    prettyAnsi (GApp x []) = fsym x
+    prettyAnsi (GApp x args) = fsym x <+> prettyargs args
+    prettyAnsi (GFormulaData s f) = text s <> align (parens (prettyAnsi f))
+    prettyAnsi (GFormulaTerm s t) = text s <> align (parens (prettyAnsi t))
+    prettyAnsi (GVar x) = prettyAnsi x
 
 instance Pretty AtomicWord where
     pretty (AtomicWord x) = (if isLowerWord x then text else squotes.text) x
 
-instance Pretty GTerm where
-    pretty (GTerm x) = pretty x
-    pretty (ColonSep x y) = pretty x <+> oper ":" <+> pretty y
-    pretty (GList xs) = let f = oper
-                        in
-                          f "[" <+> (fillSep . punctuate comma . fmap pretty) xs <+> f "]"
+instance PrettyAnsi GTerm where
+    prettyAnsi (GTerm x) = prettyAnsi x
+    prettyAnsi (ColonSep x y) = prettyAnsi x <+> oper ":" <+> prettyAnsi y
+    prettyAnsi (GList xs) = let f = oper
+                            in
+                              f "[" <+> (fillSep . punctuate comma . fmap prettyAnsi) xs <+> f "]"
 
 
 
-instance Pretty V where
-    pretty (V x) = blue . text $ x
+instance PrettyAnsi V where
+    prettyAnsi (V x) = annotate (color Blue) . text $ x

--- a/changelog.markdown
+++ b/changelog.markdown
@@ -1,3 +1,7 @@
+## 0.6.0.0
+
+* migrate to `prettyprinter` package family from `ansi-wl-pprint` package
+
 ## 0.5.1.0
 
 * Fix compilation error with `mtl >=2.3`

--- a/logic-TPTP.cabal
+++ b/logic-TPTP.cabal
@@ -1,5 +1,5 @@
 name: logic-TPTP
-version: 0.5.1.0
+version: 0.6.0.0
 cabal-version: >= 1.10
 build-type: Simple
 license: GPL

--- a/logic-TPTP.cabal
+++ b/logic-TPTP.cabal
@@ -63,7 +63,9 @@ Library
                    , array
                    , syb
                    , containers
-                   , ansi-wl-pprint < 1.0
+                   , prettyprinter >=1.2.1.1 && <1.8
+                   , prettyprinter-ansi-terminal >=1.0 && <1.2
+                   , text
                    , QuickCheck >= 2
                    , mtl
                    , pointed
@@ -96,7 +98,8 @@ Executable TestImportExportImportFile
  hs-source-dirs:    testing
  build-depends:     logic-TPTP
                   , base
-                  , ansi-wl-pprint
+                  , prettyprinter
+                  , prettyprinter-ansi-terminal
                   , optparse-applicative >=0.11 && <0.19
                   , pcre-light
                   , semigroups
@@ -114,7 +117,8 @@ Test-suite TestImportExportRandom
  hs-source-dirs:    testing
  build-depends:     logic-TPTP
                   , base
-                  , ansi-wl-pprint
+                  , prettyprinter
+                  , prettyprinter-ansi-terminal
                   , pcre-light
                   , QuickCheck
                   , semigroups
@@ -131,7 +135,8 @@ Executable PrettyPrintFile
                   , SimpleArgs
  hs-source-dirs:    testing
  build-depends:     logic-TPTP
-                  , ansi-wl-pprint
+                  , prettyprinter
+                  , prettyprinter-ansi-terminal
                   , base
                   , bytestring
                   , containers
@@ -154,7 +159,8 @@ Executable ParseRandom
                   , SimpleArgs
  hs-source-dirs:    testing
  build-depends:     logic-TPTP
-                  , ansi-wl-pprint
+                  , prettyprinter
+                  , prettyprinter-ansi-terminal
                   , base
                   , bytestring
                   , containers

--- a/testing/Common.hs
+++ b/testing/Common.hs
@@ -6,7 +6,8 @@ module Common where
 
 import Data.Monoid
 import qualified Data.Semigroup as Semigroup
-import Text.PrettyPrint.ANSI.Leijen
+import Prettyprinter
+import Prettyprinter.Render.Terminal
 import Text.Regex.PCRE.Light.Char8
 import Codec.TPTP
 
@@ -28,14 +29,14 @@ instance Monoid AFormulaComparison where
     mappend = (Semigroup.<>)
 #endif
 
-instance Pretty AFormulaComparison where
-    pretty (OtherSame) = dullgreen.text$"OtherSame"
-    pretty (OtherDiff x y) = sep
-                             [dullred.text$"OtherDiff"
+instance PrettyAnsi AFormulaComparison where
+    prettyAnsi (OtherSame) = annotate (colorDull Green).pretty$"OtherSame"
+    prettyAnsi (OtherDiff x y) = sep
+                             [annotate (colorDull Red).pretty$"OtherDiff"
                              ,pretty x
                              ,pretty y
                              ]
-    pretty (FormulaDiff fd) = pretty fd
+    prettyAnsi (FormulaDiff fd) = prettyAnsi fd
 
 compareOther ::  (Eq a, Show a) => a -> a -> AFormulaComparison
 compareOther x y = if x==y then OtherSame else OtherDiff (show x) (show y)

--- a/testing/ParseRandom.hs
+++ b/testing/ParseRandom.hs
@@ -18,7 +18,6 @@ import Debug.Trace
 import Data.Generics
 import Test.QuickCheck
 import Data.Monoid
-import Text.PrettyPrint.ANSI.Leijen
 import System.Exit
 import Text.Regex.PCRE.Light.Char8
 import Common

--- a/testing/PrettyPrintFile.hs
+++ b/testing/PrettyPrintFile.hs
@@ -19,7 +19,6 @@ import Debug.Trace
 import Data.Generics
 import Test.QuickCheck
 import Data.Monoid
-import Text.PrettyPrint.ANSI.Leijen
 import System.Exit
 import Text.Regex.PCRE.Light.Char8
 

--- a/testing/TestImportExportImportFile.hs
+++ b/testing/TestImportExportImportFile.hs
@@ -3,7 +3,8 @@ module Main where
 
 import Control.Monad
 import Data.Monoid
-import Text.PrettyPrint.ANSI.Leijen hiding ((<$>))
+import Prettyprinter
+import Prettyprinter.Render.Terminal
 import System.Exit
 import Common
 import Options.Applicative
@@ -49,7 +50,7 @@ diff_once_twice print_export print_failure_only infilename'  = do
   case findUnsupportedFormulaType input of
      Just x -> do
        unless print_failure_only $ 
-         putStrLn . prettySimple . yellow . text $ ("Skipping unsupported formula type "++x)
+         putStrLn . prettySimple . annotate (color Yellow) . text $ ("Skipping unsupported formula type "++x)
      Nothing -> do
 
       let once = parse input
@@ -58,7 +59,7 @@ diff_once_twice print_export print_failure_only infilename'  = do
         putStrLn $ "new tptp = " ++tptp
       let twice = parse tptp
       let dif = mconcat (zipWith diffAFormula once twice)
-      let success = (putStrLn . prettySimple . dullgreen . text $ "Ok")
+      let success = (putStrLn . prettySimple . annotate (colorDull Green) . text $ "Ok")
 
       -- case dif of
       --   OtherSame -> success
@@ -74,3 +75,6 @@ diff_once_twice print_export print_failure_only infilename'  = do
            when print_failure_only $ putStrLn infilename'
            putStrLn . prettySimple $ dif
            exitWith (ExitFailure 1)
+
+text :: String -> Doc ann
+text = pretty

